src/Simplex/Chat/Terminal/Input.hs

```diff
@@ -132,13 +144,18 @@ receiveFromTTY cc@ChatController {inputQ, activeTo} ct@ChatTerminal {termSize, t
         | (c == 'l' || c == 'L') && ms == ctrlKey -> submit True
         | otherwise -> update key
       _ -> update key
-    submit live =
-      atomically (readTVar termState >>= submitInput live)
-        >>= mapM_ (uncurry endLiveMessage)
-    update key = atomically $ do
-      ac <- readTVar activeTo
-      live <- isJust <$> readTVar liveMessageState
-      modifyTVar termState $ updateTermState ac live (width termSize) key
+    submit live = do
+      ts <- readTVarIO termState
+      isLive <- isJust <$> readTVarIO liveMessageState
+      when (inputString ts /= "" || isLive) $
+        atomically (submitInput live ts) >>= mapM_ (uncurry endLiveMessage)
+    update key = do
+      ac <- readTVarIO activeTo
+      live <- isJust <$> readTVarIO liveMessageState
+      ts <- readTVarIO termState
+      user_ <- readTVarIO currentUser
+      ts' <- updateTermState user_ chatStore ac live (width termSize) key ts
+      atomically $ writeTVar termState $! ts'
```

Comment:
is the only change in these functions that they're not fully atomic?

Reply:
Don’t understand the question…

Reply:
was trying to understand what's changed